zstd-safe/src/seekable.rs

```diff
@@ -0,0 +1,721 @@
+    /// A general API allowing the client to provide its own read and seek implementations.
+    ///
+    /// Initializes this seekable without having the complete compressed data in memory,
+    /// but seeks and reads `src` as required. Use this function if you are looking for
+    /// an alternative to the `ZSTD_seekable_initFile()` function.
+    #[cfg(feature = "std")]
+    #[cfg_attr(feature = "doc-cfg", doc(cfg(feature = "std")))]
+    pub fn init_advanced<F>(
+        self,
+        src: std::boxed::Box<F>,
+    ) -> Result<AdvancedSeekable<'a, F>, crate::ErrorCode>
+    where
+        F: std::io::Read + std::io::Seek,
+    {
+        let opaque = std::boxed::Box::into_raw(src) as *mut F;
+        let custom_file = zstd_sys::ZSTD_seekable_customFile {
+            opaque: opaque as *mut core::ffi::c_void,
+            read: Some(advanced_read::<F>),
+            seek: Some(advanced_seek::<F>),
+        };
+        // Safety: Just FFI
+        let code = unsafe {
+            zstd_sys::ZSTD_seekable_initAdvanced(self.0.as_ptr(), custom_file)
+        };
+
+        if crate::is_error(code) {
+            return Err(code);
+        }
+
+        Ok(AdvancedSeekable {
+            inner: self,
+            src: opaque,
+        })
+    }
+}
+
+/// Seeks the read head to `offset` from `origin`, where origin is either `SEEK_SET`
+/// (beginning of file), `SEEK_CUR` (current position) or `SEEK_END` (end of file),
+/// as defined in `stdio.h`.
+///
+/// Returns a non-negative value in case of success, and a negative value in case of failure.
+#[cfg(feature = "std")]
+unsafe extern "C" fn advanced_seek<S: std::io::Seek>(
+    opaque: *mut core::ffi::c_void,
+    offset: ::core::ffi::c_longlong,
+    origin: ::core::ffi::c_int,
+) -> ::core::ffi::c_int {
+    use core::convert::TryFrom;
+    use std::io::SeekFrom;
+
+    // as defined in stdio.h
+    const SEEK_SET: i32 = 0;
+    const SEEK_CUR: i32 = 1;
+    const SEEK_END: i32 = 2;
+
+    // Safety: The trait boundaries in `init_advanced()` ensure that `opaque` implements Seek.
+    let seeker: &mut S = std::mem::transmute(opaque);
+    let pos = match origin {
+        SEEK_SET => {
+            let Ok(offset) = u64::try_from(offset) else {
+                return -1;
+            };
+            SeekFrom::Start(offset)
+        }
+        SEEK_CUR => SeekFrom::Current(offset),
+        SEEK_END => SeekFrom::End(offset),
+        // not possible
+        _ => return -1,
+    };
+
+    if seeker.seek(pos).is_err() {
+        return -1;
+    }
+
+    0
+}
+
+/// Reads exactly `n` bytes into `buffer`.
+///
+/// Returns a non-negative value in case of success, and a negative value in case of failure.
+#[cfg(feature = "std")]
+unsafe extern "C" fn advanced_read<R: std::io::Read>(
+    opaque: *mut core::ffi::c_void,
+    buffer: *mut core::ffi::c_void,
+    n: usize,
+) -> ::core::ffi::c_int {
+    // Safety: The trait boundaries in `init_advanced()` ensure that `opaque` implements Read.
```

Comment:
Note: the safety requires that `opaque` points _exactly_ to a `R`, not just something that implements `Read`.